pkg/testing/gadgetapi/gadgetapi.go

```diff
@@ -0,0 +1,121 @@
+// Copyright 2024 The Inspektor Gadget authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package gadgetapi
+
+import (
+	"context"
+	"fmt"
+	"strings"
+	"time"
+
+	"github.com/inspektor-gadget/inspektor-gadget/pkg/operators"
+	_ "github.com/inspektor-gadget/inspektor-gadget/pkg/operators/ebpf"
+	ocihandler "github.com/inspektor-gadget/inspektor-gadget/pkg/operators/oci-handler"
+
+	"github.com/inspektor-gadget/inspektor-gadget/pkg/datasource"
+	igjson "github.com/inspektor-gadget/inspektor-gadget/pkg/datasource/formatters/json"
+	gadgetcontext "github.com/inspektor-gadget/inspektor-gadget/pkg/gadget-context"
+	"github.com/inspektor-gadget/inspektor-gadget/pkg/operators/simple"
+	"github.com/inspektor-gadget/inspektor-gadget/pkg/runtime/local"
+)
+
+type GadgetAPI struct {
+	gadgetCtx *gadgetcontext.GadgetContext
+	runtime   *local.Runtime
+	outputCh  chan string
+}
+
+// NewGadget initializes a new GadgetAPI with the specified gadget image.
+func NewGadget(image string, timeout time.Duration) (*GadgetAPI, error) {
+	const opPriority = 50000
+
+	// Create an output channel
+	outputCh := make(chan string, 100)
+
+	// Set up simple gadget operator with JSON formatting
+	gadgetOperator := simple.New("gadget", simple.OnInit(func(gadgetCtx operators.GadgetContext) error {
+		fmt.Println("Initializing gadget operator...")
+
+		for _, d := range gadgetCtx.GetDataSources() {
+			jsonFormatter, _ := igjson.New(d)
+
+			d.Subscribe(func(source datasource.DataSource, data datasource.Data) error {
+				jsonOutput := jsonFormatter.Marshal(data)
+				outputCh <- string(jsonOutput)
+				fmt.Printf("Captured data: %s\n", jsonOutput)
```

Comment:
I think we should be providing the raw `source datasource.DataSource, data datasource.Data` to the gadget. 

Perhaps `NewGadget` could have a `source datasource.DataSource, data datasource.Data` callback argument, then the each gadget can decide how to process the data, for instance, to marhsall to json.

However I'm wondering if we should go a step back, perhaps this wrapper is creating some distraction, what about moving all this logic to unit/trace_open_test.go, once we have test working, try to implement tests for few more gadgets and then see how we can create this common wrapper?

Reply:
Yeah its best to do it that way, lets get it done in future commits.